node/core/approval-voting/src/criteria.rs

```diff
@@ -483,52 +660,134 @@ pub(crate) fn check_assignment_cert(
+	{
+		if claimed_core >= config.n_cores as usize {
+			return Err(InvalidAssignment(Reason::CoreIndexOutOfBounds))
+		}
 
-	if is_in_backing {
-		return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		let is_in_backing =
+			is_in_backing_group(&config.validator_groups, validator_index, *backing_group);
+
+		if is_in_backing {
+			return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		}
 	}
 
-	let vrf_signature = &assignment.vrf;
-	match assignment.kind {
-		AssignmentCertKind::RelayVRFModulo { sample } => {
-			if sample >= config.relay_vrf_modulo_samples {
+	let vrf_output = &assignment.vrf.output;
+	let vrf_proof = &assignment.vrf.proof;
+
+	match &assignment.kind {
+		AssignmentCertKindV2::RelayVRFModuloCompact { core_bitfield } => {
+			// Check that claimed core bitfield match the one from certificate.
+			if &claimed_core_indices != core_bitfield {
+				return Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+
+			let (vrf_in_out, _) = public
+				.vrf_verify_extra(
+					relay_vrf_modulo_transcript_v2(relay_vrf_story),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_cores_transcript(core_bitfield),
+				)
+				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
+
+			let resulting_cores = relay_vrf_modulo_cores(
+				&vrf_in_out,
+				config.relay_vrf_modulo_samples,
+				config.n_cores,
+			);
+
+			// Currently validators can opt out of checking specific cores.
+			// This is the same issue to how validator can opt out and not send their assignments in the first place.
+
+			// Ensure that the `vrf_in_out` actually includes all of the claimed cores.
+			if claimed_core_indices.iter_ones().fold(true, |cores_match, core| {
+				cores_match & resulting_cores.contains(&CoreIndex(core as u32))
+			}) {
+				Ok(0)
+			} else {
+				gum::debug!(
+					target: LOG_TARGET,
+					?resulting_cores,
+					?claimed_core_indices,
+					"Assignment claimed cores mismatch",
+				);
+				Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+		},
+		AssignmentCertKindV2::RelayVRFModulo { sample } => {
+			if *sample >= config.relay_vrf_modulo_samples {
 				return Err(InvalidAssignment(Reason::SampleOutOfBounds))
 			}
 
+			// Enforce claimed candidates is 1.
+			if claimed_core_indices.count_ones() != 1 {
+				gum::warn!(
+					target: LOG_TARGET,
+					?claimed_core_indices,
+					"`RelayVRFModulo` assignment must always claim 1 core",
+				);
+				return Err(InvalidAssignment(Reason::InvalidArguments))
+			}
+
 			let (vrf_in_out, _) = public
 				.vrf_verify_extra(
-					relay_vrf_modulo_transcript(relay_vrf_story, sample),
-					&vrf_signature.output.0,
-					&vrf_signature.proof.0,
-					assigned_core_transcript(claimed_core_index),
+					relay_vrf_modulo_transcript_v1(relay_vrf_story, *sample),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_core_transcript(CoreIndex(
+						claimed_core_indices.first_one().expect("Checked above; qed") as u32,
```

Comment:
this should be extracted out to a variable as the `expect` appears multiple times.